daemon/cluster/cluster.go

```diff
@@ -334,8 +334,9 @@ func (c *Cluster) Cleanup() {
 		c.mu.Unlock()
 		return
 	}
-	defer c.mu.Unlock()
 	state := c.currentNodeState()
+	c.mu.Unlock()
```

Comment:
After removing the locks, https://github.com/docker/docker/pull/32449/files#diff-12d2a05186062a2b8f2fb8c09adac4a1R30 now applies to leave/cleanup. I think it's better to cancel handlers gracefully. That means either doing the temp lock in these places or reverting this and adding comment that we are locking `c.nr` to make sure there is nothing holding readlocks.